src/circuit/boolean.rs

```diff
@@ -375,35 +375,60 @@ impl Boolean {
         where E: Engine,
               CS: ConstraintSystem<E>
     {
-        let c = Self::xor(&mut cs, a, b)?;
+        match (a, b) {
+            // constant and a = 0 or b = 0
+            (&Boolean::Constant(false), &Boolean::Is(ref x)) |
+            (&Boolean::Constant(true), &Boolean::Not(ref x)) |
+            (&Boolean::Is(ref x), &Boolean::Constant(false)) |
+            (&Boolean::Not(ref x), &Boolean::Constant(true)) => {
+                cs.enforce(
+                    || "enforce equal",
+                    |lc| lc,
+                    |lc| lc,
+                    |lc| lc + x.get_variable()
+                );
 
-        match c {
-            Boolean::Constant(false) => {
                 Ok(())
             },
-            Boolean::Constant(true) => {
-                Err(SynthesisError::Unsatisfiable)
-            },
-            Boolean::Is(ref res) => {
+            // constant and a = 1 or b = 1
+            (&Boolean::Constant(true), &Boolean::Is(ref x)) |
+            (&Boolean::Constant(false), &Boolean::Not(ref x)) |
+            (&Boolean::Is(ref x), &Boolean::Constant(true)) |
+            (&Boolean::Not(ref x), &Boolean::Constant(false)) => {
                 cs.enforce(
-                    || "enforce equals zero",
+                    || "enforce equal",
                     |lc| lc,
                     |lc| lc,
-                    |lc| lc + res.get_variable()
+                    |lc| lc + CS::one() - x.get_variable()
                 );
 
                 Ok(())
             },
-            Boolean::Not(ref res) => {
+            // a equals not b
+            (&Boolean::Is(ref a), &Boolean::Not(ref b)) | (&Boolean::Not(ref a), &Boolean::Is(ref b)) => {
                 cs.enforce(
-                    || "enforce equals one",
-                    |lc| lc,
-                    |lc| lc,
-                    |lc| lc + CS::one() - res.get_variable()
+                    || "enforce equal",
+                    |lc| lc + a.get_variable(),
+                    |lc| lc + CS::one(),
+                    |lc| lc + CS::one() - b.get_variable()
                 );
 
                 Ok(())
             },
+            // a equals b
+            (&Boolean::Is(ref a), &Boolean::Is(ref b)) | (&Boolean::Not(ref a), &Boolean::Not(ref b)) => {
+                cs.enforce(
+                    || "enforce equal",
+                    |lc| lc + a.get_variable(),
+                    |lc| lc + CS::one(),
+                    |lc| lc + b.get_variable()
+                );
+
+                Ok(())
+            },
+            (&Boolean::Constant(_), &Boolean::Constant(_)) => {
```

Comment:
true = true and false = false are generally okay, it's just true = false / false = true that are unsatisfiable.

Reply:
Is it worth adding a test for this?